### PR TITLE
Load truffle json artifacts

### PR DESCRIPTION
**Pull request by feliam**
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/1376)

### CODE REVIEW

**Comment by disconnect3d** [comment]
Can we add a testcase for this feature?

**Comment by ggrieco-tob** [comment]
Maybe we should take a look to how `truffle` is tested?

**Comment by ggrieco-tob** [comment]
I believe we should use this schema to test this: https://github.com/trufflesuite/truffle/tree/master/packages/truffle-contract-schema

File: manticore/ethereum/manticore.py [hunk 1]
```diff
@@ -572,6 +579,76 @@ def make_symbolic_arguments(self, types):
...
+
+        if isinstance(jfile, io.IOBase):
+            jfile = jfile.read()
+        elif not isinstance(jfile, str):
+            raise TypeError(f'source code bad type: {type(jfile).__name__}')
+
+        truffle = json.loads(jfile)
+        hashes = {}
+        for item in truffle['abi']:
+            type = item['type']
```
Comment (ehennenfent) [inline manticore/ethereum/manticore.py]: Shadowing the builtin `type` here is breaking the fstrings
```suggestion
            item_type = item['type']
```

File: manticore/ethereum/manticore.py [hunk 2]
```diff
@@ -572,6 +579,76 @@ def make_symbolic_arguments(self, types):
...
+            if type in ('function'):
```
Comment (ehennenfent) [inline manticore/ethereum/manticore.py]: See [#r263540882](https://github.com/trailofbits/manticore/pull/1376/files#r263540882)
```suggestion
            if item_type in ('function'):
```

File: manticore/ethereum/manticore.py [hunk 3]
```diff
@@ -572,6 +579,76 @@ def make_symbolic_arguments(self, types):
...
+        if "deployedSourceMap" in truffle:
+            srcmap_runtime = truffle["deployedSourceMap"].split(';')
+        else:
+            srcmap_runtime = []
+        abi = truffle['abi']
+        md = SolidityMetadata(contract_name, source_code, bytecode, runtime, srcmap, srcmap_runtime, hashes, abi, b'')
+        contract_account = self.create_contract(owner=owner, balance=balance, init=md._init_bytecode)
+
+        if contract_account is None:
+            raise EthereumError("Failed to build contract %s" % contract_name_i)
```
Comment (ehennenfent) [inline manticore/ethereum/manticore.py]: Typo?
```suggestion
            raise EthereumError("Failed to build contract %s" % contract_name)
```

Comment (feliam) [reply manticore/ethereum/manticore.py]: bad copy pasta Also we need a f-string there.

File: manticore/ethereum/manticore.py [hunk 4]
```diff
@@ -323,7 +330,7 @@ def _compile(source_code, contract_name, libraries=None, solc_bin=None, solc_rem
                     break
 
         if name is None:
-            raise ValueError('Specified contract not found')
+            raise ValueError('Specified contract not found', contract_name)
```
Comment (disconnect3d) [inline manticore/ethereum/manticore.py]: Maybe:
```suggestion
            raise ValueError(f'Specified contract not found: {contract_name}')
```

File: manticore/ethereum/manticore.py [hunk 5]
```diff
@@ -572,6 +579,78 @@ def make_symbolic_arguments(self, types):
         # FIXME this is more naive than reasonable.
         return ABI.deserialize(types, self.make_symbolic_buffer(32, name='INITARGS', avoid_collisions=True))
 
+    def json_create_contract(self, jfile, owner=None, name=None, contract_name=None, balance=0, gas=None, network_id=None):
+        """ Creates a solidity contract based on a truffle json artifact.
+
+            :param jfile: truffle json artifact
```
Comment (disconnect3d) [inline manticore/ethereum/manticore.py]: Is there any chance we link or point to some source/docs regarding on how a truffle json artifact looks like?

Or to our own tests/examples.

**Comment by ESultanik** [comment]
It appears as if this new feature assumes that the contract’s constructor takes no arguments.  I am going to confirm that this morning, and then update the API to mirror the pre-existing contract creation mechanism, allowing constructor arguments.

**Comment by ESultanik** [comment]
Okay, I think that single commit fixed the constructor arguments problem.

File: manticore/ethereum/manticore.py [hunk 6]
```diff
@@ -572,6 +579,78 @@ def make_symbolic_arguments(self, types):
         # FIXME this is more naive than reasonable.
         return ABI.deserialize(types, self.make_symbolic_buffer(32, name='INITARGS', avoid_collisions=True))
 
+    def json_create_contract(self, jfile, owner=None, name=None, contract_name=None, balance=0, gas=None, network_id=None):
+        """ Creates a solidity contract based on a truffle json artifact.
+
```
Comment (ehennenfent) [inline manticore/ethereum/manticore.py]:
```suggestion
            https://github.com/trufflesuite/truffle/tree/develop/packages/truffle-contract-schema
```

File: manticore/ethereum/manticore.py (hunk 5 above)
Comment (ehennenfent) [reply manticore/ethereum/manticore.py]: https://github.com/trailofbits/manticore/pull/1376#pullrequestreview-215730566

**Pull request merged**